### PR TITLE
Open value sharing

### PR DESCRIPTION
**Pull request by lastmjs**

### CODE REVIEW

File: examples/open_value_sharing/test/test.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { getCanisterId } from 'azle/dfx';
+import { runTests } from 'azle/test';
+
+import { actorConsumer, agent } from './actor_consumer';
+import { createActor as createActorWallet } from './dfx_generated/wallet';
```
Comment (bdemann) [inline examples/open_value_sharing/test/test.ts:5]: how does createWalletActor and walletActor strike you?

**Review by bdemann — CHANGES_REQUESTED** [review]

File: src/compiler/get_consumer_config.ts [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    depth: number;
+    weight: number;
+    platform: string;
+    asset: string;
+    paymentMechanism: string;
+    custom: Record<string, any>;
+};
+
+export type ConsumerConfig = {
+    killSwitch?: Consumer['killSwitch'];
```
Comment (bdemann) [inline src/compiler/get_consumer_config.ts:39]: please add platforms and assets

File: src/compiler/generate_cargo_toml_files.ts [hunk 3]
```diff
@@ -28,7 +28,9 @@ export function generateWorkspaceCargoToml(optLevel: OptLevel): Toml {
 
         [workspace]
```
Comment (bdemann) [inline src/compiler/generate_cargo_toml_files.ts:29]: Do we even need this? Could we delete all of worspace? we seemed to be living quite happily with canister_methods this whole time

**Review by bdemann — CHANGES_REQUESTED** [review]

File: src/compiler/utils/types.ts [hunk 4]
```diff
@@ -28,6 +29,10 @@ export type JSCanisterConfig = Readonly<{
     assets_large?: [string, string][];
     esm_aliases?: Record<string, string>;
     esm_externals?: string[];
+    // TODO we should move all custom properties into custom in a subsequent PR
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/utils/types.ts:32]: Yes, and I think this should be done as part of the next release, to minimize breaking changes to the dfx.json: https://github.com/demergent-labs/azle/issues/1814

File: src/compiler/utils/get_canister_config.ts [hunk 5]
```diff
@@ -57,6 +57,7 @@ export function getCanisterConfig(
                 join('candid', 'icp', 'icrc.did')
             ],
             [
+                // TODO use require.main instead of __dirname elsewhere
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/utils/get_canister_config.ts:60]: This has not been fully resolved

File: src/compiler/rust/open_value_sharing/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    pub payments: Vec<Payment>,
+}
+
+#[derive(candid::CandidType, Clone)]
+pub struct Payment {
+    pub time: u64,
+    pub name: String,
+    pub principal: candid::Principal,
+    pub amount: u128,
+    pub success: Result<(), String>,
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/rust/open_value_sharing/src/lib.rs:30]: Let's put payment_mechanism in

File: src/compiler/rust/open_value_sharing/src/platforms/icp/assets/cycles.rs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+use crate::{get_dependency_principal, Dependency, Payment};
+
+pub async fn handle_asset_payment(
+    dependency: &Dependency,
+    amount: u128,
+) -> Result<Payment, anyhow::Error> {
+    let principal = get_dependency_principal(dependency)?;
+
+    if dependency.payment_mechanism == "wallet_receive" {
+        ic_cdk::api::call::call_with_payment128::<(Option<()>,), ()>(
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/rust/open_value_sharing/src/platforms/icp/assets/cycles.rs:12]: I'll look into what happens if a cross-canister call never returns

File: src/compiler/prepare_rust_staging_area.ts [hunk 8]
```diff
@@ -22,6 +23,11 @@ export function prepareRustStagingArea(
 
     writeFileSync(`${canisterPath}/Cargo.toml`, workspaceCargoToml);
 
+    copySync(
+        join(__dirname, '..', '..', 'Cargo.lock'),
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/prepare_rust_staging_area.ts:27]: I created a global constant AZLE_PACKAGE_PATH and have used that everywhere require.main?.path is used. I have also fixed these cases of using __dirname, but not generally

[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/prepare_rust_staging_area.ts:27]: I have now done this generally

File: src/compiler/rust/canister/src/lib.rs [hunk 9]
```diff
@@ -191,3 +193,16 @@ fn run_event_loop(context: &mut wasmedge_quickjs::Context) {
         }
     }
 }
+
+// TODO will this work for queries as well?
+#[ic_cdk_macros::update]
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/rust/canister/src/lib.rs:198]: Let's discuss this

[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/rust/canister/src/lib.rs:198]: We will not address this now, but here's an issue: https://github.com/demergent-labs/azle/issues/1816

File: examples/open_value_sharing/test/actor_consumer.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        [],
+        PeriodicBatch[]
+    >;
+}
+
+export const agent = new HttpAgent({
+    host: 'http://127.0.0.1:8000'
+});
+
+export const actorConsumer = Actor.createActor<_SERVICE>(
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply examples/open_value_sharing/test/actor_consumer.ts:33]: Fixed

File: src/compiler/rust/open_value_sharing/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/rust/open_value_sharing/src/lib.rs:30]: Done

File: src/compiler/get_consumer_config.ts [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+    const dependenciesUnnormalized =
+        await getDependenciesUnnormalized(consumerConfig);
+    const dependencies = normalizeDependencies(dependenciesUnnormalized);
+
+    const depthWeights = getDepthWeights(dependencies);
+
+    return {
+        killSwitch: consumerConfig?.killSwitch ?? DEFAULT_KILL_SWITCH,
+        sharedPercentage:
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/get_consumer_config.ts:97]: Done

File: examples/open_value_sharing/test/test.ts (hunk 1 above)
Comment (lastmjs) [reply examples/open_value_sharing/test/test.ts]: Changed

File: src/compiler/get_consumer_config.ts (hunk 2 above)
Comment (lastmjs) [reply src/compiler/get_consumer_config.ts:39]: Done

File: src/compiler/rust/open_value_sharing/src/platforms/icp/assets/cycles.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/rust/open_value_sharing/src/platforms/icp/assets/cycles.rs:12]: We must discuss the trade-offs

[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/rust/open_value_sharing/src/platforms/icp/assets/cycles.rs:12]: Let's do notify for now

File: src/compiler/generate_cargo_toml_files.ts (hunk 3 above)
Comment (lastmjs) [reply src/compiler/generate_cargo_toml_files.ts:29]: No it needs the workspace at least, so might as well put it all in?

File: src/compiler/rust/open_value_sharing/src/platforms/icp/assets/cycles.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (lastmjs) [reply src/compiler/rust/open_value_sharing/src/platforms/icp/assets/cycles.rs]: Implemented

File: src/compiler/generate_cargo_toml_files.ts (hunk 3 above)
Comment (bdemann) [reply src/compiler/generate_cargo_toml_files.ts:29]: sounds good to me

**Review by bdemann — DISMISSED** [review]

**Comment by lastmjs** [comment]
I keep getting the open value sharing tests failing with an ELSProblems

**Pull request merged**